### PR TITLE
Support Clojure 1.12

### PR DESCRIPTION
**Pull request by vincentjames501**
I was expecting to have to do more than just bump some versions but this seems to work if I install and test things locally.

This addresses https://github.com/cloverage/cloverage/issues/347 and https://github.com/cloverage/cloverage/issues/345

### CODE REVIEW

File: cloverage/project.clj [hunk 1]
```diff
@@ -17,44 +17,46 @@
...
   :deploy-repositories [["releases" :clojars]
                         ["snapshots" :clojars]]
-  :dependencies [[org.clojure/clojure "1.11.1" :scope "provided"]
-                 [org.clojure/tools.reader "1.3.6"]
-                 [org.clojure/tools.cli "0.4.2"]
-                 [org.clojure/tools.logging "0.5.0"]
-                 [org.clojure/tools.namespace "1.3.0"]
-                 [org.clojure/java.classpath "1.0.0"]
+  :dependencies [[org.clojure/clojure "1.12.0" :scope "provided"]
+                 [org.clojure/tools.reader "1.5.0"]
```
Comment (vincentjames501) [inline cloverage/project.clj:21]: I bumped other deps while I was in here. Things still seem to work OK.

File: cloverage/src/cloverage/coverage.clj [hunk 2]
```diff
@@ -23,7 +23,7 @@
            java.io.FileNotFoundException
            java.util.concurrent.atomic.AtomicInteger))
 
-(def ^:dynamic *instrumented-ns*) ;; currently instrumented ns
+(def ^:dynamic *instrumented-ns* nil) ;; currently instrumented ns
```
Comment (vincentjames501) [inline cloverage/src/cloverage/coverage.clj:26]: Fixes a lint issue

**Comment by vincentjames501** [comment]
This addresses 
https://github.com/cloverage/cloverage/issues/347
and
https://github.com/cloverage/cloverage/issues/345

**Comment by lvh** [comment]
Wonderful, thanks!

**Comment by vincentjames501** [comment]
@lvh , thanks for the quick reply. Any plans for a new release? Would love to get Clojure 1.12 support in for our teams!

**Comment by lvh** [comment]
Unfortunately it looks like CI is some type of broken.

**Comment by vincentjames501** [comment]
@lvh , oh odd. CI passed on this MR:
https://app.circleci.com/jobs/github/cloverage/cloverage/1019
https://app.circleci.com/jobs/github/cloverage/cloverage/1027

**Comment by lvh** [comment]
Yes, and bizarrely the issue is .... deploy keys?!

**Comment by lvh** [comment]
Okay, now we're getting somewhere: the problem is that the the entire CI suite doesn't run on PRs. The entire suite also runs the tests under old Clojures, which can't execute the new test, because `^String/1` is not valid syntax in pre-1.12.

**Comment by vincentjames501** [comment]
@lvh , ok it's likely this macro:

```clojure
(defmacro clj-version-or-higher
  [target-version & body]
  `(if (and (>= (compare (clojure-version) ~target-version) 0))
     ~@body
     (t/is (true? true))))
```

Since the tests passed I didn't worry about it too much. Want to try something like this?

```clojure
(defn compare-versions
  [v1 v2]
  (letfn [(parse-version [v] (mapv #(Integer/parseInt ^String %) (str/split v #"\.")))
          (pad-zeros [v1 v2]
            (let [max-length (max (count v1) (count v2))]
              [(concat v1 (repeat (- max-length (count v1)) 0))
               (concat v2 (repeat (- max-length (count v2)) 0))]))]
    (let [[v1-vec v2-vec] (pad-zeros (parse-version v1) (parse-version v2))]
     (loop [a v1-vec
            b v2-vec]
       (cond
         (empty? a) true
         (> (first a) (first b)) false
         (< (first a) (first b)) true
         :else (recur (rest a) (rest b)))))))

(defmacro clj-version-or-higher
  [target-version & body]
  `(when (compare-versions ~target-version (clojure-version))
     ~@body))

(clj-version-or-higher
 "1.12.0"
 (t/deftest instrument-clj-1-12-features
   (t/testing "Instrumentation of new Clojure 1.12 features"
     (t/testing "Qualified methods - Class/method, Class/.method, and Class/new"
       (t/is (= '(do
                   (do
                     (do ((do Long/new) (do 1)))
                     (do System/out)
                     (do (let* [f (do Long/.byteValue)]
                           (do ((do f) (do 1)))))
                     (do (let* [f (do Long/valueOf)]
                           (do ((do f) (do 1)))))))
                (rw/macroexpand-all (inst/instrument-form #'inst/nop
                                                          nil
                                                          '(do
                                                             (Long/new 1)
                                                             System/out
                                                             (let [f Long/.byteValue]
                                                               (f 1))
                                                             (let [f Long/valueOf]
                                                               (f 1))))))))
     (t/testing "Functional interfaces"
       (t/is (= '(do
                   (let* [p (do even?)]
                     (do (. p test (do 42)))))
                (rw/macroexpand-all (inst/instrument-form #'inst/nop
                                                          nil
                                                          '(let [^java.util.function.Predicate p even?]
                                                             (.test p 42)))))))
     (t/testing "Array class syntax"
       (t/is (= '(do
                   (do
                     (do (new ProcessBuilder (do ((do into-array) (do String) (do [(do "a")])))))
                     (do ((do java.util.Arrays/binarySearch)
                          (do (. clojure.lang.Numbers clojure.core/int_array (do [(do 1) (do 2) (do 3)])))
                          (do (. clojure.lang.RT (intCast (do 2))))))))
                (rw/macroexpand-all (inst/instrument-form #'inst/nop
                                                          nil
                                                          '(do
                                                             (ProcessBuilder. ^String/1 (into-array String ["a"]))
                                                             (java.util.Arrays/binarySearch ^int/1 (int-array [1 2 3])
                                                                                            (int 2)))))))))))
```

**Comment by lvh** [comment]
Yeah, I haven't tried it but I would not expect that to work because the failure is in the reader. I do think you're on to something though. I feel like we likely have 2 options, both rely on delaying reading/evaluation.

1. Put it in a string and evaluate that
2. Put it in a separate namespace that is only conditionally required. Basically: unconditionally create a test that conditionally, probably via`requiring-resolve`, loads the test body as a function

2 might mean some linters fail, but I'm OK with linters only working on 1 version.

Do you think that'd work?

**Comment by vincentjames501** [comment]
@lvh , I think that'd work and I think option 1 would be fine personally or just bump cloverage to 2.x and only support clj 1.12+. If it had failed in CI for me when making this MR I probably would've just done number 1.

**Comment by vincentjames501** [comment]
@lvh , want to give this a try?

```
(t/deftest instrument-clj-1-12-features
  (let [clj-1-12-or-higher? (or (> (:major *clojure-version*) 1)
                                (and (= (:major *clojure-version*) 1)
                                     (>= (:minor *clojure-version*) 12)))]
    (if-not clj-1-12-or-higher?
      (t/is (not clj-1-12-or-higher?) "Test skipped because Clojure version is less than 1.12")
      (t/testing "Instrumentation of new Clojure 1.12 features"
        (letfn [(instrument-form [form]
                  (rw/macroexpand-all (inst/instrument-form #'inst/nop nil (read-string form))))]
          (t/testing "Qualified methods - Class/method, Class/.method, and Class/new"
            (t/is (= '(do
                        (do
                          (do ((do Long/new) (do 1)))
                          (do System/out)
                          (do (let* [f (do Long/.byteValue)]
                                (do ((do f) (do 1)))))
                          (do (let* [f (do Long/valueOf)]
                                (do ((do f) (do 1)))))))
                     (instrument-form
                      "(do
                         (Long/new 1)
                         System/out
                         (let [f Long/.byteValue]
                           (f 1))
                         (let [f Long/valueOf]
                           (f 1)))"))))
          (t/testing "Functional interfaces"
            (t/is (= '(do
                        (let* [p (do even?)]
                          (do (. p test (do 42)))))
                     (instrument-form
                      "(let [^java.util.function.Predicate p even?]
                         (.test p 42))"))))
          (t/testing "Array class syntax"
            (t/is (= '(do
                        (do
                          (do (new ProcessBuilder (do ((do into-array) (do String) (do [(do "a")])))))
                          (do ((do java.util.Arrays/binarySearch)
                               (do (. clojure.lang.Numbers clojure.core/int_array (do [(do 1) (do 2) (do 3)])))
                               (do (. clojure.lang.RT (intCast (do 2))))))))
                     (instrument-form
                      "(do
                         (ProcessBuilder. ^String/1 (into-array String [\"a\"]))
                         (java.util.Arrays/binarySearch ^int/1 (int-array [1 2 3])
                                                        (int 2)))")))))))))
```

**Comment by vincentjames501** [comment]
@lvh , any updates? Anything I can help with?

**Pull request merged**